grpclog/loggerv2_test.go

```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+}
+
+func TestLoggerV2PrintFuncAllCombinations(t *testing.T) {
+	const (
+		print int = iota
+		printf
+		println
+	)
+	type testDiscard struct {
+		discardInf  bool
+		discardWarn bool
+		discardErr  bool
+	}
+	cfgs := []loggerV2Config{
+		{jsonFormat: false},
+		{jsonFormat: true},
+	}
+	discardTests := []testDiscard{
+		{discardInf: true, discardWarn: true, discardErr: true},
+		{discardInf: true, discardWarn: true, discardErr: false},
+		{discardInf: true, discardWarn: false, discardErr: true},
+		{discardInf: false, discardWarn: true, discardErr: true},
+		{discardInf: true, discardWarn: false, discardErr: false},
+		{discardInf: false, discardWarn: false, discardErr: true},
+		{discardInf: false, discardWarn: false, discardErr: false},
+	}
+	var printName = []string{
+		print:   "print",
+		printf:  "printf",
+		println: "println",
+	}
+	var jsonName = map[bool]string{
+		true:  "json",
+		false: "noJson",
+	}
+	discardName := func(td testDiscard) string {
+		strs := []string{}
+		if td.discardInf {
+			strs = append(strs, "discardInfo")
+		}
+		if td.discardWarn {
+			strs = append(strs, "discardWarn")
+		}
+		if td.discardErr {
+			strs = append(strs, "discardErr")
+		}
+		if len(strs) == 0 {
+			strs = append(strs, "noDiscard")
+		}
+		return strings.Join(strs, " ")
+	}
+	for _, cfg := range cfgs {
+		for printType := print; printType <= println; printType++ {
+			for _, discardTest := range discardTests {
+				testName := fmt.Sprintf("%v %v %v", jsonName[cfg.jsonFormat], printName[printType], discardName(discardTest))
+				t.Run(testName, func(t *testing.T) {
+					buffers := []*bytes.Buffer{new(bytes.Buffer), new(bytes.Buffer), new(bytes.Buffer)}
+					writers := []io.Writer{buffers[infoLog], buffers[warningLog], buffers[errorLog]}
+					if discardTest.discardInf {
+						writers[infoLog] = io.Discard
+					}
+					if discardTest.discardWarn {
+						writers[warningLog] = io.Discard
+					}
+					if discardTest.discardErr {
+						writers[errorLog] = io.Discard
+					}
+					logger := newLoggerV2WithConfig(writers[infoLog], writers[warningLog], writers[errorLog], cfg)
+
+					msgInf := "someinfo"
+					msgWarn := "somewarn"
+					msgErr := "someerr"
+					if printType == print {
+						logger.Info(msgInf)
+						logger.Warning(msgWarn)
+						logger.Error(msgErr)
+					} else if printType == printf {
+						logger.Infof("%v", msgInf)
+						logger.Warningf("%v", msgWarn)
+						logger.Errorf("%v", msgErr)
+					} else if printType == println {
+						logger.Infoln(msgInf)
+						logger.Warningln(msgWarn)
+						logger.Errorln(msgErr)
+					}
+
+					checkWrite(t, buffers[infoLog], discardTest.discardInf, "info", msgInf, cfg.jsonFormat)
+					checkWrite(t, buffers[warningLog], discardTest.discardWarn, "warning", msgWarn, cfg.jsonFormat)
+					checkWrite(t, buffers[errorLog], discardTest.discardErr, "error", msgErr, cfg.jsonFormat)
+				})
+			}
+		}
+	}
+}
+
+func TestLoggerV2Fatal(t *testing.T) {
```

Comment:
this test looks different than others. Its hard to understand what are we verifying here

Reply:
It's testing the Fatal functions. Because they call os.Exit, the only way to test them is to call them in a sub-process.
It's following https://go.dev/talks/2014/testing.slide#23

I added a comment explaining. 

I can remove them if you want. But then the Fatal functions won't be tested, and this won't have 100% test coverage.